app/dossierPrograms/dossierPrograms.controllers.js

```diff
@@ -72,66 +76,69 @@ function($scope, $translate, $anchorScroll, dossiersProgramsFactory,dossiersProg
-                if ($scope.sections.programStageSections.length == 0) $scope.showProgramWithoutSections();    
-                else $scope.showProgramWithSections();            
-            endLoadingState(true);
+            var stageSectionPromises = $scope.selectedProgram.programStages.map(function (stage) {
+                return dossiersProgramStageSectionsFactory.get({programStageId: stage.id}).$promise;
             });
+
+            $q.all(stageSectionPromises).then(function (stages) {
+                $scope.stages = stages.map(function (stage, index) {
+                    var toc = {
+                        displayName: "Stage: " + stage.displayName + (stage.repeatable ? " (Repeatable)" : ""),
+                        id: stage.id,
+                        index: index
+                    };
+                    if(stage.programStageSections.length == 0) 
+                        return createStageWithoutSections(stage, toc);
+                    else 
+                        return createStageWithSections(stage, toc);
+                });
+                endLoadingState(true);
             });
         }
     });
 
-   $scope.showProgramWithoutSections = function() {
-        $scope.stages4TOC.displayName = "Data elements";
-        //line needed to reuse code of the ng-repeat on the view
-        $scope.sections.programStageSections =  [{displayName: "Data Elements", id:"DataElements", dataElements: $scope.sections.programStageDataElements}];
-        console.log($scope.sections);
-        for (i = 0; i < $scope.sections.programStageSections[0].dataElements.length; ++i) {
-            $scope.sections.programStageSections[0].dataElements[i] = $scope.sections.programStageSections[0].dataElements[i].dataElement;
+    function createStageWithoutSections (stage, toc) {
+        // Line to make it compatible with view
+        stage.programStageSections = [{
+            displayName: "Data Elements",
+            dataElements: stage.programStageDataElements.map(function(stageDataElement) {
+                return stageDataElement.dataElement;
+            })
+        }];
+        /** 
+        for (i = 0; i < resultStage.programStageSections[0].dataElements.length; ++i) {
+            resultStage.programStageSections[0].dataElements[i] = resultStage.programStageSections[0].dataElements[i].dataElement;
         }
-        console.log("out");
-        console.log($scope.sections);
-        addtoTOC($scope.toc, null, $scope.stages4TOC, "programs");
-    };
-
-    $scope.showProgramWithSections = function() {
-        $scope.stages4TOC.displayName = "Program Sections";
-        addtoTOC($scope.toc, $scope.sections.programStageSections, $scope.stages4TOC, "programs");
+        */
+        addtoTOC($scope.toc, null, toc, "programs");
+        return stage;
     }
 
+    function createStageWithSections (stage, toc) {
+        addtoTOC($scope.toc, stage.programStageSections, toc, "programs");
+        return stage;
+    }
 
 }]);
 
 
 
-
```

Comment:
This part is intentionally commented out. Our program indicator list is too long to be displayed...